Optimized BigInt implementation

denisrosset opened the pull request:
This augments the `BigInt` class with a `Long` field, so that small integers can be handled without allocating a `java.math.BigInteger`.

The main motivation is to make `BigInt` a "safe integer" type with a reasonable performance profile in the case it is mainly used to store `Long`-sized integers; whereas the overhead of the introduced special case is dwarfed by the `java.math.BigInteger` machinery.

To maintain binary compatibility even if `BigInt` is a final class, I added a main constructor parameter of type `Long`; that main constructor is private. The original constructor is now a secondary constructor taking a single `BigInteger` argument. Logic inside the constructor decides whether to store the value as a `BigInteger` or `Long` argument.

This introduces a small memory overhead when `BigInt` stores a `BigInteger` (but in turn, `BigInteger` is a pretty mighty class which allocates an array).

Points to be discussed:

- MiMA doesn't complain: but the `val bigInteger: BigInteger` constructor argument is now a `def bigInteger: BigInteger` body method. Is that OK?

- This needs to be benchmarked. What would be a relevant benchmark in the Scala community build?

- The GCD optimization rests on a Long GCD code lifted from the Spire library, which is under the MIT License. Is such a code integration permitted by the Scala CLA?

- The Java API documentation allows for big integers with a bit length bigger than what an `Int` can store. I'm using `bitLength` liberally in the new implementation, I don't know how the code would behave in these cases. On the other hand, the old code calls `bitLength` a few times before having code to handle the "overly big BigInteger speical case". Shouldn't `bitLength` throw an exception beforehand (the API specifies that methods should throw an exception instead of returning an invalid value, but this point isn't clear). What should we do there? (It's interesting to find this type of undefined behaviour in the Java API).

denisrosset committed:
Optimized BigInt implementation

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -108,14 +109,56 @@ object BigInt {
+  //                  otherwise, _bigInteger stores the number and _long = 0L
+  def this(_bigInteger: BigInteger) = this(
+    if (_bigInteger.bitLength <= 63) null else _bigInteger,
+    if (_bigInteger.bitLength <= 63) _bigInteger.longValue else 0L
```
Are you sure bitLength is not computed twice. Isn’t it better to read it once just in case?

denisrosset replied:
I'd love to! What can I do in a secondary constructor to achieve that?

Ichoran replied:
Why even bother storing the long version if we explicitly put in a `BigInteger`?

Ichoran replied:
`private def this(big: Boolean, bi: BigInteger, l: Long) = this(if (big) bi else null, if (big) 0L else l)`

denisrosset replied:
We don't store a `BigInteger` in case the value fits in a `Long`: hence the test of the argument bit length.

This constructor has two purposes:

1. It preserves binary compatibility: `BigInt` had a primary constructor taking a `BigInteger` value. Here we store that value efficiently. Of course, a call to the `bigInteger` member is wasteful, because it reallocates a `BigInteger`. However, the purpose of the redesign is to avoid operating on `BigInteger`s whenever possible. If that constructor was not public, this logic would have gone in the companion object method `apply`, but this is not possible.

2. Sometimes, the result of an operation involving two `BigInteger`s will result in an integer that fits in a long. This method will make sure that the class invariant is preserved in that case.

denisrosset replied:
@Ichoran The different types of encoding have been clarified in the latest commit.

denisrosset replied:
@johnynek The new encoding takes care of reading `bitLength` once.

@Ichoran Happy with the clarifications?

SethTisue replied:
ping @Ichoran

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -108,14 +109,56 @@ object BigInt {
+
+  private def this(_long: Long) = this(null, _long)
+
+  def bigInteger: BigInteger = if (_bigInteger eq null) BigInteger.valueOf(_long) else _bigInteger
```
Should we cache this? If not comment why not?

denisrosset replied:
We could; though I wanted to limit the overhead of `BigInt` to an additional long. Caching this would mean 1) add another pointer-sized field to BigInt (we already have a `_bigInteger` private field due to the constructor), or 2) reuse the `_bigInteger` field, but in that case a call to `bigInteger` would deoptimize the instance behavior.

WIth this, and other comments, I'd like to get a bit of feedback about the benchmarks that I should use.

denisrosset replied:
@johnynek Actually I found a third possibility, see latest commit. Instead of using `_bigInteger eq null` as a tag meaning that the value is in the `_long`, I use `_long == Long.MinValue` as a tag that the value is in the `_bigInteger`.

And in all cases, I can use the `_bigInteger` field, now mutable, to cache the result of the `.bigInteger` call without allocating more memory.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,35 +221,85 @@ final class BigInt(val bigInteger: BigInteger)
   /** Compares this BigInt with the specified BigInt for equality.
    */
-  def equals (that: BigInt): Boolean = compare(that) == 0
+  def equals (that: BigInt): Boolean = (this eq that) || (
```
I’m skeptical the this we that pays for itself. Do you have evidence that helps?

denisrosset replied:
No. The motivation was that the `BigInt`s between -1024 and 1024 are cached. And indeed in that case we are in the happy path anyway.

denisrosset replied:
That test was removed.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -289,38 +424,56 @@ final class BigInt(val bigInteger: BigInteger)
    */
-  def testBit (n: Int): Boolean = this.bigInteger.testBit(n)
+  def testBit (n: Int): Boolean =
+    if (isValidLong && n <= 63) (_long & (1L << n)) != 0 else this.bigInteger.testBit(n)
```
Can’t we split the n >= 63 case inside isValidLong to handle it without allocating the BigInteger?

denisrosset replied:
We could, though it's additional logic due to the sign bit.

SethTisue replied:
@denisrosset decision here?

denisrosset replied:
@SethTisue Has been changed.

johnynek reviewed (commented):
This is great! Thanks for tackling this. I’m on my phone but a few questions:

Are you sure there is test coverage for all these changes? I can imagine we need to be more careful and really exercise all the branches now.

Can we commit any benchmarks to compare to BigInteger and java.lang.Long? The goal, I imagine here is be only a bit slower than a boxed Long but have the ability to scale up. Would be nice to demonstrate that in benchmarks (sorry if I overlooked them in this PR).

denisrosset commented:
Ok, MiMa complains about 
```
[error] scala-library: Failed binary compatibility check against org.scala-lang:scala-library:2.13.0! Found 1 potential problems (filtered 412)

[error]  * method this(Long)Unit in class scala.math.BigInt does not have a correspondent in other version
```
even though that (new) constructor is private. Is that a problem in practice?

Ichoran commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,35 +221,85 @@ final class BigInt(val bigInteger: BigInteger)
+    if (this.isValidLong && that.isValidLong) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long / that._long)
+    }
+    new BigInt(this.bigInteger.divide(that.bigInteger))
```
This line breaks the invariant.

Ichoran commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,35 +221,85 @@ final class BigInt(val bigInteger: BigInteger)
+    if (this.isValidLong && that.isValidLong) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long % that._long)
+    }
+    new BigInt(this.bigInteger.remainder(that.bigInteger))
```
This line also breaks the invariant.

Ichoran commented on src/library/scala/math/BigInt.scala:
```diff
@@ -380,7 +533,7 @@ final class BigInt(val bigInteger: BigInteger)
    *  `Double.POSITIVE_INFINITY` as appropriate.
    */
-  def doubleValue: Double = this.bigInteger.doubleValue
+  def doubleValue: Double = if (isValidInt) _long.toDouble else this.bigInteger.doubleValue
```
Why just `isValidInt`?  Up to `1L << 52` can be converted exactly.  Even if inexact, `toDouble` should do the same thing as the `BigInteger` version.

denisrosset replied:
FIrst of all, do we agree that the behaviour of this `doubleValue` method is unchanged, i.e. it returns the same double value as the previous `BigInt` implementation?

And yes, for sure, values up to `1L << 52` can be converted exactly. I'll change that.

sjrd replied:
Values up to `1L << 53` (not 52), and down to `-(1L << 53)`, can be converted exactly. I believe the behavior is unchanged, yes.

Moreover, unless there was something involving rounding direction that can be configured for `BigInteger`s? it should even be possible to go all the way to `if (isValidLong) _long.toDouble`.

denisrosset replied:
> it should even be possible to go all the way to if (isValidLong) _long.toDouble

"This conversion is similar to the narrowing primitive conversion from double to float as defined in section 5.1.3 of The Java™ Language Specification"

I lack the background to interpret the specification correctly; as this concerns a performance loss in a quite narrow range of cases, I'd leave it alone. What do you think?

denisrosset replied:
@sjrd Changes about the range of exact conversions have been made. What do you think outside those values? I reckon that the range that fits in a Long with more that 53 bits is pretty narrow, so I don't feel like introducing complexity there. Is this resolved?

SethTisue replied:
@sjrd @denisrosset conclusion here?

denisrosset replied:
@denisrosset The observation from @sjrd has been taken in account.

sjrd commented:
> 
> 
> Ok, MiMa complains about
> 
> ```
> [error] scala-library: Failed binary compatibility check against org.scala-lang:scala-library:2.13.0! Found 1 potential problems (filtered 412)
> 
> [error]  * method this(Long)Unit in class scala.math.BigInt does not have a correspondent in other version
> ```
> 
> even though that (new) constructor is private. Is that a problem in practice?

It's not a problem in practice. It's not source compatible (the member's not stable anymore, which means you cannot `import` its members without storing it in a `val`), but it is binary compatible.

denisrosset commented:
New encoding which caches a `BigInteger` instance after request.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -221,84 +252,83 @@ final class BigInt private (private val _bigInteger: BigInteger, private val _lo
       val y = that._long
       val z = x + y
-      if ((~(x ^ y) & (x ^ z)) >= 0L) return BigInt(z)
+      if ((~(x ^ y) & (x ^ z)) >= 0L) new BigInt(z)
```
aren't you bypassing the ability to save boxed BigInt for small values by using `new` rather than `return` here?

Aslo, I'm confused how not using return works since you don't have else branches, so, isn't this always falling through to the bottom?

If so, the whole if is a waste right? and also isn't this issue replicated below?

denisrosset replied:
Bug corrected. Will benchmark the use of the cache or not. (Noting that the original BigInt doesn't create new instances through the `apply` method, and thus doesn't use the cache)

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -221,84 +252,83 @@ final class BigInt private (private val _bigInteger: BigInteger, private val _lo
-    if (this.isValidLong && that.isValidLong) { // fast path
-      if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long % that._long)
+    if (this.longEncoding && that.longEncoding) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) new BigInt(this._long % that._long)
```
I'm not a fan of these `if` with fall through... pretty dangerous... It seems this is wrong now and the fast path is never hit.

denisrosset replied:
Good catch! I corrected those locally, will review the code in the morning before pushing the corrections.

johnynek reviewed (commented):
I like the caching, noticed something that looks like a bug.

dwijnand commented:
(We use `[nomerge]` to indicate PRs with changes that can't forward-merge.  I added the "WIP" label instead, but that practice could be redundant with draft PRs, and being able to convert PRs into draft PRs... anyways.)

denisrosset commented:
## Changes

- I've written BigInt benchmarks. They are available in a separate branch, so I can switch between implementations. See https://github.com/denisrosset/scala/tree/bigint-benchmarks

- I've addressed the comments above. For the use of cache or not, I've made a new branch, which is *not* part of this PR, but is benchmarked below.

## Help needed

- Ideas for additional relevant benchmarks? Is it micro-op benchmarking worthwhile (i.e. test method by method)?

- Is there any way to automate running a bunch of JMH tests and collect the results?

- The standard deviation is all over the place; I'm running tests on a laptop (Ubuntu 20.04), from the command line. Any advice?

I'm using the following JMH annotations. `@BenchmarkMode(Array(Mode.AverageTime)) @Fork(2) @Threads(1) @Warmup(iterations = 10) @Measurement(iterations = 10) @OutputTimeUnit(TimeUnit.NANOSECONDS) @State(Scope.Benchmark)`

- Any advice to reduce the feedback cycle time? Right now, one benchmark takes around 10 minutes to run which is not ideal to iterate.

- Conversely, any advice for the parameters for the final comparison between options? I.e. evidence that I would present in this PR? I can afford then to wait a bit more.

## Preliminary benchmark results

I'm comparing:

1. the original 2.13 BigInt implementation in https://github.com/scala/scala/commit/b499fcf8cdbeecf7fa8e5454ff182fcd9f7e6775

2. this PR https://github.com/denisrosset/scala/commit/0f9c5f90634c346347425eb64b2f41a9784226fb

3. a variant that uses `BigInt.apply` and thus the cache https://github.com/denisrosset/scala/commit/60f69d80b4544de38efbeaa4dfde02ab71671702

For the [RSA benchmark](https://github.com/denisrosset/scala/blob/bigint-benchmarks/test/benchmarks/src/main/scala/scala/math/BigIntRSABenchmark.scala), I get:

1. 8564.509 ±(99.9%) 293.898 ns/op [Average]
2. 8543.954 ±(99.9%) 38.654 ns/op [Average]
3. 8689.300 ±(99.9%) 143.148 ns/op [Average]

For the [factorial benchmark](https://github.com/denisrosset/scala/blob/bigint-benchmarks/test/benchmarks/src/main/scala/scala/math/BigIntFactorialBenchmark.scala):

![FactorialBenchmark](https://user-images.githubusercontent.com/93707/80538779-4da39800-8974-11ea-89ff-8c67fa2a26db.png)

Data of the factorial benchmark below:

## 1. Latest 2.13.x library
```
Benchmark                           (size)  Mode  Cnt     Score    Error  Units
BigIntFactorialBenchmark.factorial       5  avgt   20    91.325 ±  0.783  ns/op
BigIntFactorialBenchmark.factorial      10  avgt   20   189.208 ±  0.942  ns/op
BigIntFactorialBenchmark.factorial      15  avgt   20   286.799 ±  3.006  ns/op
BigIntFactorialBenchmark.factorial      20  avgt   20   400.816 ±  5.075  ns/op
BigIntFactorialBenchmark.factorial      25  avgt   20   520.375 ± 13.439  ns/op
BigIntFactorialBenchmark.factorial      30  avgt   20   651.065 ±  5.901  ns/op
BigIntFactorialBenchmark.factorial      35  avgt   20   756.722 ± 14.020  ns/op
BigIntFactorialBenchmark.factorial      40  avgt   20   908.575 ± 15.089  ns/op
BigIntFactorialBenchmark.factorial      45  avgt   20  1034.421 ± 20.981  ns/op
BigIntFactorialBenchmark.factorial      50  avgt   20  1182.123 ± 14.954  ns/op
BigIntFactorialBenchmark.factorial      55  avgt   20  1309.791 ±  8.394  ns/op
BigIntFactorialBenchmark.factorial      60  avgt   20  1432.263 ± 24.768  ns/op
BigIntFactorialBenchmark.factorial      65  avgt   20  1553.675 ± 13.574  ns/op
BigIntFactorialBenchmark.factorial      70  avgt   20  1702.169 ± 11.971  ns/op
BigIntFactorialBenchmark.factorial      75  avgt   20  1889.631 ± 24.950  ns/op
BigIntFactorialBenchmark.factorial      80  avgt   20  2027.780 ± 26.827  ns/op
BigIntFactorialBenchmark.factorial      85  avgt   20  2226.690 ± 40.736  ns/op
BigIntFactorialBenchmark.factorial      90  avgt   20  2423.215 ± 39.091  ns/op
BigIntFactorialBenchmark.factorial      95  avgt   20  2587.199 ± 51.026  ns/op
BigIntFactorialBenchmark.factorial     100  avgt   20  2755.073 ± 40.325  ns/op
```

## 2. Optimized BigInt without cache
```
Benchmark                           (size)  Mode  Cnt     Score    Error  Units
BigIntFactorialBenchmark.factorial       5  avgt   20    51.850 ±  0.719  ns/op
BigIntFactorialBenchmark.factorial      10  avgt   20    97.546 ±  1.250  ns/op
BigIntFactorialBenchmark.factorial      15  avgt   20   142.426 ±  2.590  ns/op
BigIntFactorialBenchmark.factorial      20  avgt   20   186.384 ±  2.220  ns/op
BigIntFactorialBenchmark.factorial      25  avgt   20   377.740 ±  4.264  ns/op
BigIntFactorialBenchmark.factorial      30  avgt   20   536.705 ±  7.151  ns/op
BigIntFactorialBenchmark.factorial      35  avgt   20   675.759 ±  9.551  ns/op
BigIntFactorialBenchmark.factorial      40  avgt   20   839.809 ±  4.489  ns/op
BigIntFactorialBenchmark.factorial      45  avgt   20   955.144 ± 30.492  ns/op
BigIntFactorialBenchmark.factorial      50  avgt   20  1080.100 ± 11.415  ns/op
BigIntFactorialBenchmark.factorial      55  avgt   20  1201.680 ± 21.882  ns/op
BigIntFactorialBenchmark.factorial      60  avgt   20  1429.358 ± 18.297  ns/op
BigIntFactorialBenchmark.factorial      65  avgt   20  1548.369 ± 10.125  ns/op
BigIntFactorialBenchmark.factorial      70  avgt   20  1711.400 ± 21.597  ns/op
BigIntFactorialBenchmark.factorial      75  avgt   20  1947.580 ± 70.605  ns/op
BigIntFactorialBenchmark.factorial      80  avgt   20  2016.584 ± 35.509  ns/op
BigIntFactorialBenchmark.factorial      85  avgt   20  2198.158 ± 12.886  ns/op
BigIntFactorialBenchmark.factorial      90  avgt   20  2454.478 ± 67.294  ns/op
BigIntFactorialBenchmark.factorial      95  avgt   20  2703.225 ± 83.176  ns/op
BigIntFactorialBenchmark.factorial     100  avgt   20  2875.966 ± 87.845  ns/op
```
## 3. Optimized BigInt with cache
```
Benchmark                           (size)  Mode  Cnt     Score    Error  Units
BigIntFactorialBenchmark.factorial       5  avgt   20    50.208 ±   1.377  ns/op
BigIntFactorialBenchmark.factorial      10  avgt   20   105.166 ±   1.669  ns/op
BigIntFactorialBenchmark.factorial      15  avgt   20   154.791 ±   2.989  ns/op
BigIntFactorialBenchmark.factorial      20  avgt   20   202.301 ±   2.199  ns/op
BigIntFactorialBenchmark.factorial      25  avgt   20   389.317 ±   5.932  ns/op
BigIntFactorialBenchmark.factorial      30  avgt   20   529.150 ±   6.468  ns/op
BigIntFactorialBenchmark.factorial      35  avgt   20   687.965 ±  15.501  ns/op
BigIntFactorialBenchmark.factorial      40  avgt   20   832.886 ±  15.126  ns/op
BigIntFactorialBenchmark.factorial      45  avgt   20  1004.644 ±  18.778  ns/op
BigIntFactorialBenchmark.factorial      50  avgt   20  1148.712 ±  25.158  ns/op
BigIntFactorialBenchmark.factorial      55  avgt   20  1240.980 ±  19.981  ns/op
BigIntFactorialBenchmark.factorial      60  avgt   20  1475.280 ±  60.774  ns/op
BigIntFactorialBenchmark.factorial      65  avgt   20  1687.045 ±  55.694  ns/op
BigIntFactorialBenchmark.factorial      70  avgt   20  1744.787 ±  44.528  ns/op
BigIntFactorialBenchmark.factorial      75  avgt   20  2004.772 ±  52.582  ns/op
BigIntFactorialBenchmark.factorial      80  avgt   20  2154.762 ±  76.210  ns/op
BigIntFactorialBenchmark.factorial      85  avgt   20  2381.689 ±  66.024  ns/op
BigIntFactorialBenchmark.factorial      90  avgt   20  2510.976 ±  37.217  ns/op
BigIntFactorialBenchmark.factorial      95  avgt   20  2810.998 ±  60.709  ns/op
BigIntFactorialBenchmark.factorial     100  avgt   20  2976.557 ± 113.094  ns/op
```

denisrosset commented:
Caching in `BigInt.apply` was calling the wrong constructor. After correction, the version using the cache is faster.

```
Latest 2.13.x library
Benchmark                           (size)  Mode  Cnt     Score    Error  Units
BigIntFactorialBenchmark.factorial       5  avgt   20    91.325 ±  0.783  ns/op
BigIntFactorialBenchmark.factorial      10  avgt   20   189.208 ±  0.942  ns/op
BigIntFactorialBenchmark.factorial      15  avgt   20   286.799 ±  3.006  ns/op
BigIntFactorialBenchmark.factorial      20  avgt   20   400.816 ±  5.075  ns/op
BigIntFactorialBenchmark.factorial      25  avgt   20   520.375 ± 13.439  ns/op
BigIntFactorialBenchmark.factorial      30  avgt   20   651.065 ±  5.901  ns/op
BigIntFactorialBenchmark.factorial      35  avgt   20   756.722 ± 14.020  ns/op
BigIntFactorialBenchmark.factorial      40  avgt   20   908.575 ± 15.089  ns/op
BigIntFactorialBenchmark.factorial      45  avgt   20  1034.421 ± 20.981  ns/op
BigIntFactorialBenchmark.factorial      50  avgt   20  1182.123 ± 14.954  ns/op
BigIntFactorialBenchmark.factorial      55  avgt   20  1309.791 ±  8.394  ns/op
BigIntFactorialBenchmark.factorial      60  avgt   20  1432.263 ± 24.768  ns/op
BigIntFactorialBenchmark.factorial      65  avgt   20  1553.675 ± 13.574  ns/op
BigIntFactorialBenchmark.factorial      70  avgt   20  1702.169 ± 11.971  ns/op
BigIntFactorialBenchmark.factorial      75  avgt   20  1889.631 ± 24.950  ns/op
BigIntFactorialBenchmark.factorial      80  avgt   20  2027.780 ± 26.827  ns/op
BigIntFactorialBenchmark.factorial      85  avgt   20  2226.690 ± 40.736  ns/op
BigIntFactorialBenchmark.factorial      90  avgt   20  2423.215 ± 39.091  ns/op
BigIntFactorialBenchmark.factorial      95  avgt   20  2587.199 ± 51.026  ns/op
BigIntFactorialBenchmark.factorial     100  avgt   20  2755.073 ± 40.325  ns/op

Optimized BigInt without cache
Benchmark                           (size)  Mode  Cnt     Score    Error  Units
BigIntFactorialBenchmark.factorial       5  avgt   20    56.005 ±  2.844  ns/op
BigIntFactorialBenchmark.factorial      10  avgt   20    99.144 ±  0.638  ns/op
BigIntFactorialBenchmark.factorial      15  avgt   20   146.020 ±  0.859  ns/op
BigIntFactorialBenchmark.factorial      20  avgt   20   190.778 ±  2.982  ns/op
BigIntFactorialBenchmark.factorial      25  avgt   20   371.584 ±  6.375  ns/op
BigIntFactorialBenchmark.factorial      30  avgt   20   536.971 ±  5.861  ns/op
BigIntFactorialBenchmark.factorial      35  avgt   20   668.471 ±  7.294  ns/op
BigIntFactorialBenchmark.factorial      40  avgt   20   842.652 ± 11.786  ns/op
BigIntFactorialBenchmark.factorial      45  avgt   20   941.192 ± 12.239  ns/op
BigIntFactorialBenchmark.factorial      50  avgt   20  1099.328 ± 10.535  ns/op
BigIntFactorialBenchmark.factorial      55  avgt   20  1250.473 ± 19.429  ns/op
BigIntFactorialBenchmark.factorial      60  avgt   20  1402.944 ± 34.437  ns/op
BigIntFactorialBenchmark.factorial      65  avgt   20  1553.778 ± 23.653  ns/op
BigIntFactorialBenchmark.factorial      70  avgt   20  1670.973 ± 23.180  ns/op
BigIntFactorialBenchmark.factorial      75  avgt   20  1915.087 ± 54.379  ns/op
BigIntFactorialBenchmark.factorial      80  avgt   20  2048.740 ± 47.212  ns/op
BigIntFactorialBenchmark.factorial      85  avgt   20  2217.776 ± 23.843  ns/op
BigIntFactorialBenchmark.factorial      90  avgt   20  2397.801 ± 43.543  ns/op
BigIntFactorialBenchmark.factorial      95  avgt   20  2621.795 ± 40.232  ns/op
BigIntFactorialBenchmark.factorial     100  avgt   20  2828.747 ± 58.761  ns/op

Optimized BigInt with cache
Benchmark                           (size)  Mode  Cnt     Score    Error  Units
BigIntFactorialBenchmark.factorial       5  avgt   20    47.680 ±   0.416  ns/op
BigIntFactorialBenchmark.factorial      10  avgt   20    91.619 ±   0.746  ns/op
BigIntFactorialBenchmark.factorial      15  avgt   20   152.791 ±   0.571  ns/op
BigIntFactorialBenchmark.factorial      20  avgt   20   205.353 ±   2.425  ns/op
BigIntFactorialBenchmark.factorial      25  avgt   20   388.793 ±  13.057  ns/op
BigIntFactorialBenchmark.factorial      30  avgt   20   520.664 ±   9.426  ns/op
BigIntFactorialBenchmark.factorial      35  avgt   20   693.699 ±  18.214  ns/op
BigIntFactorialBenchmark.factorial      40  avgt   20   835.729 ±  40.293  ns/op
BigIntFactorialBenchmark.factorial      45  avgt   20   988.396 ±  12.616  ns/op
BigIntFactorialBenchmark.factorial      50  avgt   20  1099.307 ±  17.729  ns/op
BigIntFactorialBenchmark.factorial      55  avgt   20  1208.516 ±  43.706  ns/op
BigIntFactorialBenchmark.factorial      60  avgt   20  1416.259 ±  20.404  ns/op
BigIntFactorialBenchmark.factorial      65  avgt   20  1548.893 ±  42.810  ns/op
BigIntFactorialBenchmark.factorial      70  avgt   20  1750.144 ±  28.317  ns/op
BigIntFactorialBenchmark.factorial      75  avgt   20  1873.494 ±  34.342  ns/op
BigIntFactorialBenchmark.factorial      80  avgt   20  2050.726 ±  17.714  ns/op
BigIntFactorialBenchmark.factorial      85  avgt   20  2250.785 ±  28.706  ns/op
BigIntFactorialBenchmark.factorial      90  avgt   20  2485.059 ± 130.190  ns/op
BigIntFactorialBenchmark.factorial      95  avgt   20  2594.462 ±  48.606  ns/op
BigIntFactorialBenchmark.factorial     100  avgt   20  2845.801 ±  46.668  ns/op
```

lrytz commented:
>     * Is there any way to automate running a bunch of JMH tests and collect the results?

Mabye https://github.com/lightbend/benchdb could be helpful? It's very new.

>     * The standard deviation is all over the place; I'm running tests on a laptop (Ubuntu 20.04), from the command line. Any advice?

Most errors seem to be around 1%, which seems reasonable no?

https://github.com/scala/scala-dev/issues/338 has information, probably more than you need :-) Generally, I could imagine results being less stable on laptops because they usually don't have enough cooling to keep the CPU frequency stable.

>     * Any advice to reduce the feedback cycle time? Right now, one benchmark takes around 10 minutes to run which is not ideal to iterate.

Hmm, 10 minutes is not that much for a benchmark, but obviously annoying if you can't hand it off to a different machine. In fact you might actually want to do more than 2 forks if you need more results. Did you check (manually, by looking at the numbers) that 10 warmup iterations are enough?

Maybe you can do a log scale on the sizes that you measure and thin it out.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -108,14 +116,85 @@ object BigInt {
+   *  Long.MinValue. */
+  private def longEncoding: Boolean = _long != Long.MinValue
+
+  def bigInteger: BigInteger =
```
since `_bigInteger` is a var, it might be slightly clearer to write:
```scala
def bigInteger: BigInteger = {
  val read = _bigInteger
  if (read ne null) read
  else {
    val write = BigInteger.valueOf(_long)
    _bigInteger = write
    write
  }
}

```
In this way, we don't access a mutable variable twice. on the happy path.

javax-swing replied:
why not just use a lazy val here?

denisrosset replied:
@johnynek Understood

@javax-swing: In essence, the `_bigInteger` field implements a lazy val, and our encoding avoids the CPU overhead of locking, and the memory overhead of a lazy field (4 bytes for the BigInteger reference, plus the lazy val boolean bitmap).

SethTisue replied:
@denisrosset should this conversation be marked as resolved?

denisrosset replied:
Yes, resolved.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
-  def equals (that: BigInt): Boolean = compare(that) == 0
+  def equals (that: BigInt): Boolean =
+    if (this.longEncoding) {
+      if (that.longEncoding) this._long == that._long else false
```
`that.longEncoding && (this._long == that.long)`
reads a bit clearer if you ask me, but this is a nitpick.

denisrosset replied:
Changed to
```scala
    if (this.longEncoding)
      that.longEncoding && (this._long == that._long)
    else
      !that.longEncoding && (this._bigInteger == that._bigInteger)
```

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
+      val x = this._long
+      val y = that._long
+      val z = x + y
+      if ((~(x ^ y) & (x ^ z)) >= 0L) return new BigInt(z)
```
can you comment why `new BigInt(z)` here and not `return BigInt(z)` which could take advantage of cached boxed values?

denisrosset replied:
For sure!

The original code, for some reason, did not use the cache for unary or binary operations. I wonder if there was any reason (conditional tests being expensive?). I made two variants, the one using the cache is here: https://github.com/denisrosset/scala/commits/bigint-cached

In the factorial benchmark above, which doesn't hit the cache a lot but doesn't stress the memory either, there is basically no speed difference. I'll be merging the caching changes.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
+      val x = this._long
+      val y = that._long
+      val z = x - y
+      if (((x ^ y) & (x ^ z)) >= 0L) return new BigInt(z)
```
same request for a comment about cached boxed values.

denisrosset replied:
Same as above, the variant was benchmarked.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
+      val x = this._long
+      val y = that._long
+      val z = x * y
+      if (x == 0 || (y == z / x && !(x == -1 && y == Long.MinValue))) return new BigInt(z)
```
comment why not `BigInt.apply(z)`

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
-  def /  (that: BigInt): BigInt = new BigInt(this.bigInteger.divide(that.bigInteger))
+  def /  (that: BigInt): BigInt = {
+    if (this.longEncoding && that.longEncoding) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) return new BigInt(this._long / that._long)
```
comment about why not `BigInt.apply` for caching.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
-  def %  (that: BigInt): BigInt = new BigInt(this.bigInteger.remainder(that.bigInteger))
+  def %  (that: BigInt): BigInt = {
+    if (this.longEncoding && that.longEncoding) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) return new BigInt(this._long % that._long)
```
cache?

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
+    if (this.longEncoding && that.longEncoding) {
+      val x = this._long
+      val y = that._long
+      if (x != Long.MinValue || y != -1) return (new BigInt(x / y), new BigInt(x % y))
```
cache?

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
-
+  def >> (n: Int): BigInt =
+    if (longEncoding && n >= 0) {
+      if (n < 64) new BigInt(_long >> n)
```
why not `BigInt(_long >> n)` here, seems like we would often be making it small enough to hit our cache.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
-  def &  (that: BigInt): BigInt = new BigInt(this.bigInteger.and(that.bigInteger))
+  def &  (that: BigInt): BigInt =
+    if (this.longEncoding && that.longEncoding)
+      new BigInt(this._long & that._long)
```
cache?

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
+      if (that.longEncoding) {
+        if (that._long == 0) return this.abs
+        // if (that._long == Long.MinValue) return this gcd (-that)
+        new BigInt(BigInt.longGcd(this._long.abs, that._long.abs))
```
why not check the cache here? gcd is often low, seems like we could often get lucky and find things in the cache.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -176,80 +257,174 @@ final class BigInt(val bigInteger: BigInteger)
-  def mod (that: BigInt): BigInt = new BigInt(this.bigInteger.mod(that.bigInteger))
+  def mod (that: BigInt): BigInt =
+    if (this.longEncoding && that.longEncoding) {
+      val res = this._long % that._long
```
if `that._long` is `< maxCached` this is always going to be in the cache, right? Seems worth it to check to avoid an allocation.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -267,18 +442,19 @@ final class BigInt(val bigInteger: BigInteger)
    */
-  def unary_- : BigInt   = new BigInt(this.bigInteger.negate())
+  def unary_- : BigInt   =
+    if (longEncoding) new BigInt(-_long) else new BigInt(this.bigInteger.negate())
```
if we are in the cache, `- x` will generally be in the cache too. Should we not check?

denisrosset replied:
Same as above; but I won't cherry pick which paths use the caching `.apply` and which paths use a fresh instance, as I don't have a good idea of the performance profile of user code using `BigInt`.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -289,38 +465,61 @@ final class BigInt(val bigInteger: BigInteger)
    */
-  def flipBit (n: Int): BigInt  = new BigInt(this.bigInteger.flipBit(n))
+  def flipBit (n: Int): BigInt  = // note that we do not operate on the Long sign bit #63
+    if (longEncoding && n <= 62) new BigInt(_long ^ (1L << n)) else new BigInt(this.bigInteger.flipBit(n))
```
another case for the cache potentially.

denisrosset replied:
Yes!

johnynek reviewed (commented):
I made the same comment over and over: why are we doing `new BigInt(long)` rather than `BigInt.apply(long)` and seeing if we can find it in the cache.

denisrosset commented:
@lrytz 

> Mabye https://github.com/lightbend/benchdb could be helpful? It's very new.

So new, that I think I can't access it! Is the project still private?

> Most errors seem to be around 1%, which seems reasonable no?

I'm seeing the std dev magnitude fluctuating a lot between benchmark instances, so I was wondering if it comes from JIT instability, or CPU performance instability (throttling etc)

> [scala/scala-dev#338](https://github.com/scala/scala-dev/issues/338) has information, probably more than you need :-) Generally, I could imagine results being less stable on laptops because they usually don't have enough cooling to keep the CPU frequency stable.

I'll move the benchmarks to a dedicated machine, and will use two tips: disable HT and Turbo Boost.

> Hmm, 10 minutes is not that much for a benchmark, but obviously annoying if you can't hand it off to a different machine. In fact you might actually want to do more than 2 forks if you need more results. Did you check (manually, by looking at the numbers) that 10 warmup iterations are enough?

I can reduce the number of test cases now. How many forks do people usually ask for? I'm cargo culting the JMH parameters for now.

I guess to decide on the number of warmup iterations, I should set a large number like 50, and then check manually the convergence (average, std dev) in the output?

And also perform a larger number of forks, and compare the behaviour across forks?

And it's time to write a small tool to interpret/plot the benchmark results (I'm cut and pasting things in OpenOffice for now).

linasm commented:
> I'm seeing the std dev magnitude fluctuating a lot between benchmark instances, so I was wondering if it comes from JIT instability, or CPU performance instability (throttling etc)

When benchmarking on a laptop, make sure you have Turbo Boost disabled:
https://askubuntu.com/questions/619875/disabling-intel-turbo-boost-in-ubuntu

lrytz commented:
> So new, that I think I can't access it! Is the project still private?

oops 😄 i'll let you know!

> it's time to write a small tool to interpret/plot the benchmark results

that's exactly what benchdb would do

> How many forks do people usually ask for?
> check manually the convergence (average, std dev) in the output?

That's what I'd do, yeah. I personally don't have a lot of experience with bechmarking and JMH. Maybe @retronym can debunk some of the bad advice I'm giving here.

> performance profile in the benchmark will depend on whether the benchmark runs until V2 triggers GC, right?

Yes, I think that's right. Quantifying the benefit of reducing the allocation rate of an implementation is extremely difficult because the effect on GC is unpredictable. There can even be unexpected side-effects, like a lot of short-lived allocation can be cheaper (handled by minor GCs) compared to a long-lived cache that gets promoted to old gen.

I assume JMH would do its best to force a GC before starting to measure a benchmark? I don't know.

There's also the option to enable a profiler while running benchmarks https://hg.openjdk.java.net/code-tools/jmh/file/b6f87aa2a687/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_35_Profilers.java#l52, `-prof gc` should give some insight.

lrytz commented:
> > Mabye [lightbend/benchdb](https://github.com/lightbend/benchdb) could be helpful? It's very new.
> 
> So new, that I think I can't access it! Is the project still private?

The benchdb repository is public now, please take a look and feel free to give it a try! It was developed and used by @szeiger while he was working on the new Vector implementation that landed in 2.13.2 (https://github.com/scala/scala/pull/8534).

johnynek commented:
Denis: thanks for the reply about when to use `new BigInt` vs `BigInt.apply`. I would imagine though that the calculus is different now. Since the Long wrapping versions are also much more likely to be smaller values, and comparing Long to an Int (all that’s needed for checking if it is cached) is so cheap, my guess is that for the Long returning branches using caching will often be a win for code that is using many smaller values.

dwijnand commented:
@denisrosset could you rebase and squash the commits, please?  Other than that would you say this is ready for merge?

SethTisue commented:
/rebuild

SethTisue commented:
about making benchmarking available to contributors so they don't have to always run it on their own machines, there's a ticket on that now, https://github.com/scala/scala-dev/issues/753, that Jason is looking into

SethTisue commented:
Where this currently stands, afaict:

* This is nearly ready for merge for 2.13.5. We can do this!
* @denisrosset there are some conversations above not yet marked "resolved", can you check on those?
* @denisrosset earlier you mentioned that you had benchmarking code in a separate branch — is that all part of the PR now?
* @denisrosset/all was the "bit length bigger than what an Int can store" question answered? is it a blocker?
* I will check on the license question about bringing in code from Spire

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -110,14 +117,87 @@ object BigInt {
+
+  def bigInteger: BigInteger = {
+    val read = _bigInteger
+    if (read ne null) _bigInteger else {
```
why not use `if (read ne null) read` so that we don't have to read a field twice potentially?

NthPortal replied:
in fact, I don't believe it is thread-safe as it is now

viktorklang replied:
@NthPortal According to JLS it is: https://docs.oracle.com/javase/specs/jls/se15/html/jls-17.html#jls-17.7

I believe the strategy here is to avoid having coordinated updates to the reference and instead allow different cores to observe different instances assigned to `_bigInteger`. In essence, this would mean that two threads (A & B) could call `bigInteger` and pass those obtained instances to eachother via some means, and would be able to observe that they are in fact not the same instance (yet they should be equal).

I do agree that returning `read` makes a bunch more sense than reading the field twice.

This is just my 2c tho.

NthPortal replied:
@viktorklang my understanding (which I admit may be flawed) is that, since another thread's write to `_bigInteger` is racy (that is, there is no HB relationship between this thread's read and the other thread's write), each read from `_bigInteger` in this thread is permitted to view the value of `_bigInteger` either before (`null`) or after the other thread's write, and it's not clear to me that the first read in this thread being non-`null` requires the second read to be as well.

viktorklang replied:
@NthPortal Indeed. reading more than once should be a problem in general, as the method gives no guarantee as to the stability of the identity of the _biginteger's value. But reading more than once would be uncessesary and slower than just returning the value in the local variable which should be in a register already.

denisrosset replied:
Corrected!

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -291,38 +467,61 @@ final class BigInt(val bigInteger: BigInteger)
   /** Returns true if and only if the designated bit is set.
    */
-  def testBit (n: Int): Boolean = this.bigInteger.testBit(n)
+  def testBit (n: Int): Boolean =
```
are there laws that these functions all behave the same as BigInteger? I think those would be good laws to have.

SethTisue replied:
we do have ScalaCheck in this repo, so yeah, I agree some property-based tests should be added — @denisrosset unless you looked and found that they already existed

NthPortal replied:
can probably also be copied from Spire

denisrosset replied:
There is a ScalaCheck property for testBit, setBit, clearBit, flipBit etc.

plokhotnyuk commented on src/library/scala/math/BigInt.scala:
```diff
@@ -178,80 +260,173 @@ final class BigInt(val bigInteger: BigInteger)
+      val x = this._long
+      val y = that._long
+      val z = x * y
+      if (x == 0 || (y == z / x && !(x == -1 && y == Long.MinValue))) return BigInt(z)
```
Division is too expensive operation and can took ~100 cycles on contemporary CPUs.
Better option would be using of `Math.multiplyHigh` or its non-intrinstic implementation:
```
val x1 = x >> 32
val x2 = x & 0xFFFFFFFFL
val y1 = y >> 32
val y2 = y & 0xFFFFFFFFL
val z2 = x2 * y2
val t = x1 * y2 + (z2 >>> 32)
val z1 = x2 * y1 + (t & 0xFFFFFFFFL)
x1 * y1 + (t >> 32) + (z1 >> 32) != 0
```

denisrosset commented:
@SethTisue (and all): I'm currently leaving Canada for Europe and flying in ~2 hours. I'll have a deep look in 2-3 days, if that works for everyone.

SethTisue commented:
Now that the holidays have passed, I'm pinging again — perhaps folks have more time now to take a second look. (We might build 2.13.5 as soon as early February.)

NthPortal commented on src/library/scala/math/BigInt.scala:
```diff
@@ -178,80 +260,173 @@ final class BigInt(val bigInteger: BigInteger)
    */
-  def /  (that: BigInt): BigInt = new BigInt(this.bigInteger.divide(that.bigInteger))
+  def /  (that: BigInt): BigInt = {
+    if (this.longEncoding && that.longEncoding) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long / that._long)
+    }
+    new BigInt(this.bigInteger.divide(that.bigInteger))
```
`longEncoding = _long != Long.MinValue`, so the first part of this condition is always true. so you can simplify this a bit

```suggestion
    if (this.longEncoding && that.longEncoding) BigInt(this._long / that._long) // fast path
    else new BigInt(this.bigInteger.divide(that.bigInteger))
```

NthPortal commented on src/library/scala/math/BigInt.scala:
```diff
@@ -178,80 +260,173 @@ final class BigInt(val bigInteger: BigInteger)
   /** Division of BigInts
    */
-  def /  (that: BigInt): BigInt = new BigInt(this.bigInteger.divide(that.bigInteger))
+  def /  (that: BigInt): BigInt = {
+    if (this.longEncoding && that.longEncoding) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long / that._long)
+    }
+    new BigInt(this.bigInteger.divide(that.bigInteger))
+  }
```
`longEncoding = _long != Long.MinValue`, so the first part of this condition is always true. so you can simplify this a bit

```suggestion
  def /  (that: BigInt): BigInt =
    if (this.longEncoding && that.longEncoding) BigInt(this._long / that._long) // fast path
    else new BigInt(this.bigInteger.divide(that.bigInteger))
```

johnynek replied:
I don't think your suggestion is safe. If you have `Long.MinValue / -1L` you will get Long.MinValue, which is not correct:

```
scala> Long.MinValue / -1L
res1: Long = -9223372036854775808
```

The main question I have with this code, is what is the test coverage of this change. I don't think it should be merged without 100% for something so important. I'm not totally sure we have tested everything clearly. Having a scala check that BigInt is isomorphic to BigInteger for all public operations would be what I hope we can get to.

NthPortal replied:
please check the definition of `longEncoding` in the code, and see that it is precisely as I stated. thus, within the block for `if (this.longEncoding && ...)`, it must be the case that `this._long != Long.MinValue`. in which case checking that same condition again is redundant

NthPortal replied:
we could certainly add a comment noting that `this._long != Long.MinValue` and therefore no other checks are needed, for clarity

johnynek replied:
Indeed, you are correct, I was mistaken. I agree the comment would help future readers not make my same mistake.

denisrosset replied:
I've simplified the code in the direction suggested, and put the original code as a comment.

NthPortal commented on src/library/scala/math/BigInt.scala:
```diff
@@ -178,80 +260,173 @@ final class BigInt(val bigInteger: BigInteger)
   /** Remainder of BigInts
    */
-  def %  (that: BigInt): BigInt = new BigInt(this.bigInteger.remainder(that.bigInteger))
+  def %  (that: BigInt): BigInt = {
+    if (this.longEncoding && that.longEncoding) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long % that._long)
+    }
+    new BigInt(this.bigInteger.remainder(that.bigInteger))
+  }
```
same as `/`

```suggestion
  def %  (that: BigInt): BigInt =
    if (this.longEncoding && that.longEncoding) BigInt(this._long % that._long) // fast path
    else new BigInt(this.bigInteger.remainder(that.bigInteger))
```

johnynek replied:
this one may be safe because anything mod -1 is zero I think.

NthPortal commented on src/library/scala/math/BigInt.scala:
```diff
@@ -178,80 +260,173 @@ final class BigInt(val bigInteger: BigInteger)
 
   /** Returns a pair of two BigInts containing (this / that) and (this % that).
    */
   def /% (that: BigInt): (BigInt, BigInt) = {
+    if (this.longEncoding && that.longEncoding) {
+      val x = this._long
+      val y = that._long
+      if (x != Long.MinValue || y != -1) return (BigInt(x / y), BigInt(x % y))
+    }
     val dr = this.bigInteger.divideAndRemainder(that.bigInteger)
     (new BigInt(dr(0)), new BigInt(dr(1)))
   }
```
same as `/`

```suggestion
  def /% (that: BigInt): (BigInt, BigInt) =
    if (this.longEncoding && that.longEncoding) {
      val x = this._long
      val y = that._long
      (BigInt(x / y), BigInt(x % y))
    } else {
      val dr = this.bigInteger.divideAndRemainder(that.bigInteger)
      (new BigInt(dr(0)), new BigInt(dr(1)))
    }
```

johnynek replied:
this change isn't safe due to the issue I mentioned above.

denisrosset replied:
Done

NthPortal commented on src/library/scala/math/BigInt.scala:
```diff
@@ -382,7 +581,9 @@ final class BigInt(val bigInteger: BigInteger)
    */
-  def doubleValue: Double = this.bigInteger.doubleValue
+  def doubleValue: Double =
+    if (isValidLong && (-(1L << 53) <= _long && _long <= (1L << 53))) _long.toDouble
```
is it worth reading `_long` into a local variable so it only gets read once, like you did in `/%`?

denisrosset replied:
The field is not volatile; I assume this will be optimized by the JVM. Note that /% calls longEncoding (as a lot of other methods), and we'll encounter the same problem there. I'd err in the direction of readability when in doubt.

NthPortal commented:
I intend to have another pass at reviewing this—my comments so far were just my initial look (which looked good!)

SethTisue commented:
@NthPortal still planning on returning to this?

NthPortal commented on src/library/scala/math/BigInt.scala:
```diff
@@ -110,14 +117,87 @@ object BigInt {
+  // Additionally, we know that if this.isValidLong is true, then _long is the encoded value.
+
+  /** Public constructor present for compatibility. Use the BigInt.apply companion object method instead. */
+  def this(bi: BigInteger) = this(
```
should probably keep the name the same as it was previously, in case someone (for some reason) was doing `new BigInt(bigInteger = BigInteger.ZERO)`

denisrosset replied:
Added a test to check that compatibility.

NthPortal commented on src/library/scala/math/BigInt.scala:
```diff
@@ -132,11 +212,13 @@ final class BigInt(val bigInteger: BigInteger)
-  override def isValidInt: Boolean = this >= Int.MinValue && this <= Int.MaxValue
-           def isValidLong: Boolean = this >= Long.MinValue && this <= Long.MaxValue
+
+  override def isValidByte: Boolean = isValidLong && _long >= Byte.MinValue && _long <= Byte.MaxValue
+  override def isValidShort: Boolean = isValidLong && _long >= Short.MinValue && _long <= Short.MaxValue
+  override def isValidChar: Boolean = isValidLong && _long >= Char.MinValue && _long <= Char.MaxValue
+  override def isValidInt: Boolean = isValidLong && _long >= Int.MinValue && _long <= Int.MaxValue
```
because `Long.MinValue` does not fit into any of these types, we can probably use `longEncoding` instead of `isValidLong` for these 4 methods

```suggestion
  override def isValidByte: Boolean  = longEncoding && _long >= Byte.MinValue && _long <= Byte.MaxValue
  override def isValidShort: Boolean = longEncoding && _long >= Short.MinValue && _long <= Short.MaxValue
  override def isValidChar: Boolean  = longEncoding && _long >= Char.MinValue && _long <= Char.MaxValue
  override def isValidInt: Boolean   = longEncoding && _long >= Int.MinValue && _long <= Int.MaxValue
```

denisrosset replied:
Done!

NthPortal commented on src/library/scala/math/BigInt.scala:
```diff
@@ -178,80 +260,173 @@ final class BigInt(val bigInteger: BigInteger)
+      val x = this._long
+      val y = that._long
+      val z = x * y
+      if (x == 0 || (y == z / x && !(x == -1 && y == Long.MinValue))) return BigInt(z)
```
I honestly don't understand any of these checks, so I'm just going to have to assume they're right. tests that check for long overflow/underflow would be great

denisrosset replied:
We basically perform the multiplication, with possible overflow. Then we have two cases:
- either `x == 0`, then `z == 0` which is a valid `Long`,
- otherwise, we test that `y == z/x`. If `z` did overflow, the result `z/x` would be smaller than `y`.

In that last test, there could be a pathological case when `y = Long.MinValue`, but that case never happens when we have `.longEncoding`.

NthPortal commented:
@SethTisue sorry, I kinda forgot.

I ran out of steam by bit shifting, so I have no comments on that or anything after. for everything before that in the file, if I didn't comment on it, LGTM

SethTisue commented:
@denisrosset can I interest you in making one more pass over the unresolved conversations/issues above? it seems like this is nearly ready for merge

SethTisue commented:
as for further benchmarking, I suggest that be considered future work. clearly this is a win overall

SethTisue commented:
@denisrosset ping: still out there, interested?

denisrosset committed:
Merge remote-tracking branch 'scala/2.13.x' into bigint-optimization

denisrosset committed:
Modifications according to comments

dwijnand commented:
Happy Birthday, #8932!  Let's get this merged...

denisrosset committed:
Merge branch '2.13.x' into bigint-optimization

denisrosset committed:
Optimized check in BigInt./

denisrosset committed:
BigInt.% optimization

denisrosset committed:
new BigInt: preserve constructor syntax

denisrosset committed:
BigInt.isValid{Byte,Short,Char,Int} optimization

denisrosset commented:
@SethTisue 

As far as I see, we had the following concerns:

>>    @Ichoran were all your concerns addressed?

Haven't seen a reply, but I don't see any blockers.

>>  @denisrosset there are some conversations above not yet marked "resolved", can you check on those?

I've looked at all conversations, and made the changes.

>> @denisrosset earlier you mentioned that you had benchmarking code in a separate branch — is that all part of the PR now?
>> as for further benchmarking of details, I suggest that be considered possible-future-work. clearly this is a win overall

I did not merge benchmarking code; it's been a while! I'd be happy to have a second look at it later.

>>    @denisrosset / all: was the "bit length bigger than what an Int can store" question answered? is it a blocker?

>>    I will check on the license question about bringing in code from Spire
>>        (licenses are compatible, I just need to make sure we have the right notices in the right places)

Are we OK?

>> No other concerns that I can see. This is nearly ready for merge for 2.13.5. We can do this!

Agreed!

I'd be happy to glance at the code coverage before we merge, though; or have somebody review the code after the `/%` method (where @NthPortal stopped their review).

denisrosset committed:
BigInt.* clarification/optimization

denisrosset committed:
BigInt./% optimization

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -110,14 +117,87 @@ object BigInt {
+ *
+ * It wraps `java.math.BigInteger`, with optimization for small values that can be encoded in a `Long`.
+ */
+final class   BigInt private (private var _bigInteger: BigInteger, private val _long: Long)
```
looks like extra space after `class`

denisrosset replied:
Corrected

johnynek commented on project/MimaFilters.scala:
```diff
@@ -25,6 +25,10 @@ object MimaFilters extends AutoPlugin {
 
+    // for the method this(Long)Unit in class scala.math.BigInt does not have a correspondent in other version
+    // which is nevertheless private
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.math.BigInt.this"),
```
is this allowed? I thought we needed forward and backward compatibility?

denisrosset replied:
The new constructor is private; I added a comment to explain that. (The primary constructor changes and is now private, and the former public primary constructor became a public secondary constructor).  I'll let you resolve the point if you're satisfied.

denisrosset replied:
@johnynek I've removed this constructor, the MiMa compatibility filter should no longer be needed.

Edit: the MiMa filter is still required, albeit it covers only the new private constructor.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -47,8 +55,7 @@ object BigInt {
   def apply(l: Long): BigInt =
-    if (minCached <= l && l <= maxCached) apply(l.toInt)
-    else new BigInt(BigInteger.valueOf(l))
+    if (minCached <= l && l <= maxCached) getCached(l.toInt) else new BigInt(l)
```
does it pay to do `val i = l.toInt` first? I don't know if the comparisons are converting to int twice behind the scenes.

denisrosset replied:
It's more that the minCached, maxCached will be converted to Long. Truncating the Long and then doing the comparison would lead to incorrect results -- or we'd need to check `l.isValidInt`.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -110,14 +117,87 @@ object BigInt {
+      Long.MinValue
+  )
+
+  private def this(_long: Long) = this(if (_long == Long.MinValue) BigInt.longMinValueAsBigInteger else null, _long)
```
do we need this? Can't we just use `BigInt.apply` and put the logic there? This would remove the mima issue, no?

denisrosset replied:
Indeed!... especially as BigInt.apply has the caching logic. I've made all BigInt construction pass by the apply method. ~This solves the MiMa issue as well.~ No, this doesn't solve the mima check (it did on my machine).

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -132,11 +212,13 @@ final class BigInt(val bigInteger: BigInteger)
+  override def isValidShort: Boolean = longEncoding && _long >= Short.MinValue && _long <= Short.MaxValue
+  override def isValidChar: Boolean = longEncoding && _long >= Char.MinValue && _long <= Char.MaxValue
+  override def isValidInt: Boolean = longEncoding && _long >= Int.MinValue && _long <= Int.MaxValue
+           def isValidLong: Boolean = longEncoding || _bigInteger.bitLength() <= 63 // rhs of || tests == Long.MinValue
```
why not more direct: `longEncoding || (_bigInteger == BigInt.longMinValueAsBigInteger)`

denisrosset replied:
BigInteger.equals and BigInteger.bigLength are roughly comparable, complexity-wise; but the current call is slightly safe in the case we encounter a spurious `new BigInt(BigInteger.valueOf(1), Long.MinValue)` that violates the class invariants. Note that such spurious BigInt were constructed before; I've made sure that all BigInt constructions pass by the apply method, which enforces this now.

johnynek replied:
I'm very skeptical of being defensive against invariant violations. It's very hard to reason about what else may go wrong in such cases. I think we should test well enough, and encapsulate strongly enough that we are very very sure the invariants are not violated.

SethTisue replied:
@denisrosset resolution here? (it doesn't seem like a big deal to me either way)

denisrosset replied:
Changed in the direction of @johnynek suggestion

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -178,80 +260,182 @@ final class BigInt(val bigInteger: BigInteger)
+      // original code to avoid storing -Long.MinValue in a long:
+      // if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long / that._long)
+      // but we know this._long cannot be Long.MinValue, because Long.MinValue is the marker of the use of BigInteger
+      return BigInt(this._long / that._long)
```
we don't need `return` here right? we can do:
```scala
if (...) BigInt(this._long / that._long)
else new BigInt(...)
```

denisrosset replied:
Corrected, and in other places as well.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -178,80 +260,182 @@ final class BigInt(val bigInteger: BigInteger)
+      // original code to avoid storing -Long.MinValue in a long:
+      // if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long % that._long)
+      // but we know this._long cannot be Long.MinValue, because Long.MinValue is the marker of the use of BigInteger
+      return BigInt(this._long % that._long)
```
no need for return, let's use if/else

denisrosset replied:
Thanks!

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -178,80 +260,182 @@ final class BigInt(val bigInteger: BigInteger)
+      val x = this._long
+      val y = that._long
+      // original line: if (x != Long.MinValue || y != -1) return (BigInt(x / y), BigInt(x % y))
+      return (BigInt(x / y), BigInt(x % y))
```
no need for return, prefer if/else

denisrosset replied:
Thanks!

johnynek reviewed (commented):
a few more comments.

Thanks a lot for pushing on this. It would be really amazing to improve the performance of BigInt.

denisrosset committed:
Space after class

denisrosset committed:
Added comment for the Mima exclusion of the new BigInt constructor

denisrosset committed:
Removed BigInt(Long) constructor

All BigInt creations pass by the BigInt.apply method

Disabled implicit conversions in BigInt own's code

Introduced BigInt.bigTag as an alias of Long.MinValue

Removed unnecessary return statements

denisrosset committed:
Documented BigInt.bigTag

denisrosset commented:
@SethTisue 

As far as I see, we had the following concerns:

>>    @Ichoran were all your concerns addressed?

Haven't seen a reply, but I don't see any blockers.

>>  @denisrosset there are some conversations above not yet marked "resolved", can you check on those?

I've looked at all conversations, and made the changes.

>> @denisrosset earlier you mentioned that you had benchmarking code in a separate branch — is that all part of the PR now?
>> as for further benchmarking of details, I suggest that be considered possible-future-work. clearly this is a win overall

I did not merge benchmarking code; it's been a while! I'd be happy to have a second look at it later.

>>    @denisrosset / all: was the "bit length bigger than what an Int can store" question answered? is it a blocker?

>>    I will check on the license question about bringing in code from Spire
>>        (licenses are compatible, I just need to make sure we have the right notices in the right places)

Are we OK?

>> No other concerns that I can see. This is nearly ready for merge for 2.13.5. We can do this!

Agreed!

****

Code coverage-wise (by the ScalaCheck test):

- hashCode: the "big" case isn't checked, but it calls `bigInteger.##` as before,
- isValidFloat/isValidDouble: the case for `bitLen > 24` / `bitLen > 53` isn't checked, but the code is the same as before,
- the until/to methods aren't covered, but they haven't changed,
- I've introduced dummy private implicit conversion methods to deactivate implicit conversions in the BigInt code itself, and of course they aren't called.

For the testing, are we confident that the BigInt code is exercised by external code (crypto? math?)?

denisrosset committed:
Adding @unused annotation

denisrosset committed:
Put the MiMa filter back

denisrosset commented:
We still need the MiMa filter, as we add a new (private) constructor to a class.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -14,16 +14,30 @@ package scala
 
 object BigInt {
 
+  private final val bigTag = java.lang.Long.MIN_VALUE
```
why do you use MIN_VALUE here, but Long.MinValue below? Is it because you are leveraging inlining of `final val` without type? If so, maybe comment so no one changes it later. Even so, is it somehow better to use that vs `Long.MinValue`?

denisrosset replied:
I tried to make the tag scheme explicit, but it actually muddies the waters. Reverted

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -360,15 +567,15 @@ final class BigInt(val bigInteger: BigInteger)
    *  the opposite sign.
    */
-  def intValue: Int = this.bigInteger.intValue
+  def intValue: Int = if (longEncoding) _long.intValue else this.bigInteger.intValue
```
why `_long.intValue` here vs `_long.toInt`?

denisrosset replied:
Good catch! this would allocate a java.lang.Long

johnynek reviewed (commented):
Just a couple more comments.

SethTisue commented:
@denisrosset we may build 2.13.6 as soon as Mon May 17, so there's still a chance this could make it in. care to respond to Oscar's latest review round?

SethTisue commented:
> For the testing, are we confident that the BigInt code is exercised by external code (crypto? math?)?

There is presumably some usage in the community build, which will run after the PR is merged.

The license question is still on my plate; I'll get to it this week.

johnynek commented on src/library/scala/math/BigInt.scala:
```diff
@@ -132,11 +222,13 @@ final class BigInt(val bigInteger: BigInteger)
+  override def isValidByte: Boolean = longEncoding && _long >= Byte.MinValue && _long <= Byte.MaxValue
+  override def isValidShort: Boolean = longEncoding && _long >= Short.MinValue && _long <= Short.MaxValue
+  override def isValidChar: Boolean = longEncoding && _long >= Char.MinValue && _long <= Char.MaxValue
+  override def isValidInt: Boolean = longEncoding && _long >= Int.MinValue && _long <= Int.MaxValue
```
we don't need `longEncoding` on the first 4, do we? Because that code is just going to compare the long to MinValue. Why no just directly do the comparisons?

denisrosset replied:
Good catch!

denisrosset committed:
Taking in account latest round of comments

denisrosset committed:
Merge branch '2.13.x' into bigint-optimization

denisrosset committed:
Replace Spire gcd by Guava gcd algorithm (Guava is under Apache 2.0)

denisrosset committed:
Removed [this] qualifier on BigInt.longGcd (was a bug)

denisrosset committed:
Removed duplicate private implicits to disable conversions (used during code elaboration)

denisrosset commented:
@johnynek @SethTisue The code in this PR and #9628 are now exactly the same, except that #9628 has a cleaner history which should make things easier to review.

denisrosset committed:
Added comment about nonnegativity on BigInt.longGcd

SethTisue commented:
Let's merge #968 on Thursday unless there is further review feedback.
